README.md
# Ghost Cloudflare R2 Storage Adapter
[Cloudflare R2](https://www.cloudflare.com/products/r2/) storage adapter for [Ghost](https://github.com/TryGhost/Ghost).

## Features
- Save images in Cloudflare R2
- Supports images, media and files
- Resize images to emulate [Responsive Images](https://ghost.org/docs/themes/assets/)
  - Implements `saveRaw` to force ghost to generate the `srcset` attribute for image cards
- Save images and media using UUID as name
- Compress resized images
- Written in TypeScript for maintainability

## Installation
The adapter can be installed using npm or Docker.

### Clone repo
```bash
# Starting from the Ghost base directory
mkdir -p content/adapters/storage
cd content/adapters/storage
git clone https://github.com/egeldenhuys/ghost-cloudflare-r2
cd ghost-cloudflare-r2
git checkout v0.1.0
```

### Install using npm
Requires `npm` to be installed.

```bash
npm install
cp -f ./build/src/index.js ./build/src/index.js.map ./build/src/index.d.ts .
```

### Install using Docker
Requires Docker to be installed. This has the advantage of not requiring you to pollute your server with node and npm if you are already using Docker.

```bash
./build-in-docker.sh
```

## Configuration
The storage adapter makes use of the following environment variables:

| Environment Variable                               | Description                                                                                                                                                                                         |
|----------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `GHOST_STORAGE_ADAPTER_R2_ENDPOINT`                | Cloudflare R2 Endpoint. Example: `https://<account_id>.r2.cloudflarestorage.com`                                                                                                                    |
| `GHOST_STORAGE_ADAPTER_R2_ACCESS_KEY_ID`           | Access Key ID from Cloudflare R2 API Token                                                                                                                                                          |
| `GHOST_STORAGE_ADAPTER_R2_SECRET_ACCESS_KEY`       | Secret Access Key from Cloudflare R2 API Token                                                                                                                                                      |
| `GHOST_STORAGE_ADAPTER_R2_BUCKET`                  | R2 Bucket to use for storage                                                                                                                                                                        |
| `GHOST_STORAGE_ADAPTER_R2_DOMAIN`                  | R2 Custom domain to use for serving content                                                                                                                                                         |
| `GHOST_STORAGE_ADAPTER_R2_UUID_NAME`               | Use UUID as name when storing images. May cause issues when used with Responsive Images. Default `false`. Allowed values `true`, `false`                                                            |
| `GHOST_STORAGE_ADAPTER_R2_IMAGES_URL_PREFIX`       | URL prefix to use for storing and serving images from R2. Default `/content/images/`                                                                                                                |
| `GHOST_STORAGE_ADAPTER_R2_MEDIA_URL_PREFIX`        | URL prefix to use for storing and serving media (video) from R2. Default `/content/media/`                                                                                                          |
| `GHOST_STORAGE_ADAPTER_R2_FILES_URL_PREFIX`        | URL prefix to use for storing and serving files from R2. Default `/content/files/`                                                                                                                  |
| `GHOST_STORAGE_ADAPTER_R2_CONTENT_PREFIX`          | Prefix to apply to all prefixes. Default empty. Must not contain a trailing slash. Example `/blog_data`                                                                                             |
| `GHOST_STORAGE_ADAPTER_R2_GHOST_RESIZE`            | This needs to be set to `false` if Image resizing is disabled for Ghost ( env `imageOptimization__resize`). Default `true`                                                                          |
| `GHOST_STORAGE_ADAPTER_R2_RESPONSIVE_IMAGES`       | Generate an image for each width specified. Uses undocumented Ghost internal logic to get `srcset` generated. Default `false`. Allowed values `true`, `false`                                       |
| `GHOST_STORAGE_ADAPTER_R2_SAVE_ORIGINAL`           | Save the original unoptimized image. Only applicable if (env `imageOptimization__resize`) is set. Default `true`. Allowed Values `true`, `false`                                                    |
| `GHOST_STORAGE_ADAPTER_R2_RESIZE_WIDTHS`           | Comma separated list of widths to resize the image when saving. This should match the `srcset` of your theme and any Ghost overrides. Default `300,600,1000,1600,400,750,960,1140,1200`             |
| `GHOST_STORAGE_ADAPTER_R2_RESIZE_JPEG_QUALITY`     | Quality to use when resizing JPEG images. Default: `80`                                                                                                                                             |
| `GHOST_STORAGE_ADAPTER_R2_LOG_LEVEL`               | Log level for the storage adapter. Default `info`. Allowed values `debug`, `info`, `warn`, `error`                                                                                                  |
| `GHOST_STORAGE_ADAPTER_R2_SAVE_ORIG_NAME_METADATA` | Save the original file name in the object Metadata under the key `original_name`. Useful for correlating original images to images with UUID names. Default `false`. Allowed values `true`, `false` |

All environment variables can also be used as keys in the JSON config.
The following Ghost configuration is required to activate the plugin for `images`, `media`, and `files`:
Alternatively they can be specified as environment variables (See docker-compose example below).
```json
"storage": {
  "active": "ghost-cloudflare-r2",
  "ghost-cloudflare-r2": {
    "GHOST_STORAGE_ADAPTER_R2_ENDPOINT": "https://<account_id>.r2.cloudflarestorage.com"
    ...
  },
  "media": {
    "adapter": "ghost-cloudflare-r2",
    "storage_type_media": true
  },
  "files": {
    "adapter": "ghost-cloudflare-r2",
    "storage_type_files": true
  }
}
```

The section for `media` and `files` and be removed if the adapter should not handle those types.
Note: this is an undocumented syntax and might change in future Ghost releases (tested on 5.30.1).
See [Configuring Storage Adapters](https://ghost.org/docs/config/#storage-adapters) for more details.

### Redirects for backwards compatibility
If your blog is already live, and you have sent out newsletters with images, then you no longer have control over the image URLs in the emails.
The URLs will be pointing to `example.com/content/images/*` but you want to serve them from the CDN `cdn.example.com/content/images/*`.

One solution is to use [Ghost Redirects](https://ghost.org/tutorials/implementing-redirects/) (assuming your content has been copied to the CDN):
```yaml
# Temporary redirect if you might be changing the CDN in the future
302:
  ^\/content\/images\/(.*)$: https://cdn.example.com/content/images/$1
  ^\/content\/media\/(.*)$: https://cdn.example.com/content/media/$1
  ^\/content\/files\/(.*)$: https://cdn.example.com/content/files/$1
```

If you want the flexibility to later change to a different blog domain or CDN you can set
`GHOST_STORAGE_ADAPTER_R2_DOMAIN` to `blog.example.com` and use Ghost to redirect old and new image requests to the CDN.

### Example Docker Compose environment variables
```yaml
environment:
  storage__active: ghost-cloudflare-r2
  storage__media__adapter: ghost-cloudflare-r2
  storage__media__storage_type_media: true
  storage__files__adapter: ghost-cloudflare-r2
  storage__files__storage_type_files: true
  GHOST_STORAGE_ADAPTER_R2_ENDPOINT: https://<account_id>.r2.cloudflarestorage.com
  GHOST_STORAGE_ADAPTER_R2_ACCESS_KEY_ID: xxxxxx
  GHOST_STORAGE_ADAPTER_R2_SECRET_ACCESS_KEY: xxxxxx
  GHOST_STORAGE_ADAPTER_R2_BUCKET: my-ghost-bucket
  GHOST_STORAGE_ADAPTER_R2_DOMAIN: https://cdn.example.com
  GHOST_STORAGE_ADAPTER_R2_UUID_NAME: false  # optional. Default: false
  GHOST_STORAGE_ADAPTER_R2_IMAGES_URL_PREFIX: /content/images/  # optional. Default: /content/images/
  GHOST_STORAGE_ADAPTER_R2_MEDIA_URL_PREFIX: /content/media/  # optional. Default: /content/media/
  GHOST_STORAGE_ADAPTER_R2_FILES_URL_PREFIX: /content/files/  # optional. Default: /content/files/
  GHOST_STORAGE_ADAPTER_R2_CONTENT_PREFIX: ''  # optional. Default: ''
  GHOST_STORAGE_ADAPTER_R2_GHOST_RESIZE: true  # optional. Default: true
  GHOST_STORAGE_ADAPTER_R2_RESPONSIVE_IMAGES: false  # optional. Default: false
  GHOST_STORAGE_ADAPTER_R2_SAVE_ORIGINAL: true  # optional. Default: true
  # Example widths to get Dawn theme working correctly:
  GHOST_STORAGE_ADAPTER_R2_RESIZE_WIDTHS: 300,600,1000,1600,400,750,960,1140,1200 # optional. Default: 300,600,1000,1600,400,750,960,1140,1200
  GHOST_STORAGE_ADAPTER_R2_RESIZE_JPEG_QUALITY: 80  # optional. Default: 80
  GHOST_STORAGE_ADAPTER_R2_LOG_LEVEL: info  # optional. Default: info
  GHOST_STORAGE_ADAPTER_R2_SAVE_ORIG_NAME_METADATA: false  # optional. Default: false
```

## Testing
The tests require a S3 compatible endpoint. A docker-compose file for [MinIO](https://min.io/) has been included to run
a local instance for testing.

Also note that the tests will generate random images in `/tmp`.

The tests can be refactored and improved.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0] - 2023-01-30

### Added
- Option to save the original file name in the object Metadata under the key `original_name`.

### Fixed
- Fixed image generation for testing

## [0.0.1] - 2023-01-29

Initial Release

[0.1.0]: https://github.com/egeldenhuys/ghost-cloudflare-r2/releases/tag/v0.0.1...v0.1.0
[0.0.1]: https://github.com/egeldenhuys/ghost-cloudflare-r2/releases/tag/v0.0.1


src/index.ts
import StorageBase from 'ghost-storage-base';
import {Handler} from 'express-serve-static-core';
import pkg from 'loglevel';

const {getLogger} = pkg;
import {Logger} from 'loglevel';
import sharp from 'sharp';
import {v4 as uuidv4} from 'uuid';
import mime from 'mime';

const log = getLogger('ghost-cloudflare-r2');
setLogLevel(log, 'GHOST_STORAGE_ADAPTER_R2_LOG_LEVEL');

import {
  S3Client,
  PutObjectCommand,
  HeadObjectCommand,
  GetObjectCommand,
} from '@aws-sdk/client-s3';
import path from 'path';
import {readFile} from 'fs';

export interface FileInfo extends StorageBase.Image {
  originalname?: string;
  targetDir?: string;
  newPath?: string;
  originalPath?: string;
  fieldname?: string;
  encoding?: string;
  mimetype?: string;
  destination?: string;
  filename?: string;
  size?: number;
  ext?: string;
  // type is undefined when being imported
  // type?: string;
}

function stripLeadingSlash(s: string): string {
  return s.indexOf('/') === 0 ? s.substring(1) : s;
}

function stripEndingSlash(s: string): string {
  return s.indexOf('/') === s.length - 1 ? s.substring(0, s.length - 1) : s;
}

function readFileAsync(filePath: string): Promise<Buffer> {
  return new Promise((resolve, reject) =>
    readFile(filePath, (err, data) => (err ? reject(err) : resolve(data)))
  );
}

function setLogLevel(logger: Logger, envVariable: string) {
  switch (process.env[envVariable] || '') {
    case 'trace':
      logger.setLevel('trace');
      break;
    case 'debug':
      logger.setLevel('debug');
      break;
    case 'info':
      logger.setLevel('info');
      break;
    case 'warn':
      logger.setLevel('warn');
      break;
    case 'error':
      logger.setLevel('error');
      break;
    default:
      logger.setLevel('info');
  }
}

enum StorageType {
  Images = 0,
  Media = 1,
  Files = 2,
}

interface Config {
  storage_type_images?: boolean;
  storage_type_media?: boolean;
  storage_type_files?: boolean;
  GHOST_STORAGE_ADAPTER_R2_DOMAIN?: string;
  GHOST_STORAGE_ADAPTER_R2_BUCKET?: string;
  GHOST_STORAGE_ADAPTER_R2_ENDPOINT?: string;
  GHOST_STORAGE_ADAPTER_R2_ACCESS_KEY_ID?: string;
  GHOST_STORAGE_ADAPTER_R2_SECRET_ACCESS_KEY?: string;
  GHOST_STORAGE_ADAPTER_R2_IMAGES_URL_PREFIX?: string;
  GHOST_STORAGE_ADAPTER_R2_MEDIA_URL_PREFIX?: string;
  GHOST_STORAGE_ADAPTER_R2_FILES_URL_PREFIX?: string;
  GHOST_STORAGE_ADAPTER_R2_CONTENT_PREFIX?: string;
  GHOST_STORAGE_ADAPTER_R2_RESPONSIVE_IMAGES?: boolean;
  GHOST_STORAGE_ADAPTER_R2_RESIZE_WIDTHS?: string;
  GHOST_STORAGE_ADAPTER_R2_UUID_NAME?: boolean;
  GHOST_STORAGE_ADAPTER_R2_RESIZE_JPEG_QUALITY?: number;
  GHOST_STORAGE_ADAPTER_R2_SAVE_ORIGINAL?: boolean;
  GHOST_STORAGE_ADAPTER_R2_GHOST_RESIZE?: boolean;
  GHOST_STORAGE_ADAPTER_R2_SAVE_ORIG_NAME_METADATA?: boolean;
}

function getBooleanFromEnv(envName: string): boolean | undefined {
  let result: boolean | undefined;
  if (process.env[envName] && process.env[envName] === 'true') {
    result = true;
  } else if (process.env[envName] && process.env[envName] === 'false') {
    result = false;
  } else if (process.env[envName]) {
    throw new Error(
      `Environment variable ${envName} contains invalid value ${process.env[envName]}`
    );
  }

  return result;
}

function mergeConfigWithEnv(config: Config): Config {
  config.GHOST_STORAGE_ADAPTER_R2_DOMAIN =
    process.env.GHOST_STORAGE_ADAPTER_R2_DOMAIN ||
    config.GHOST_STORAGE_ADAPTER_R2_DOMAIN;

  config.GHOST_STORAGE_ADAPTER_R2_BUCKET =
    process.env.GHOST_STORAGE_ADAPTER_R2_BUCKET ||
    config.GHOST_STORAGE_ADAPTER_R2_BUCKET;

  config.GHOST_STORAGE_ADAPTER_R2_ENDPOINT =
    process.env.GHOST_STORAGE_ADAPTER_R2_ENDPOINT ||
    config.GHOST_STORAGE_ADAPTER_R2_ENDPOINT;

  config.GHOST_STORAGE_ADAPTER_R2_ACCESS_KEY_ID =
    process.env.GHOST_STORAGE_ADAPTER_R2_ACCESS_KEY_ID ||
    config.GHOST_STORAGE_ADAPTER_R2_ACCESS_KEY_ID;

  config.GHOST_STORAGE_ADAPTER_R2_SECRET_ACCESS_KEY =
    process.env.GHOST_STORAGE_ADAPTER_R2_SECRET_ACCESS_KEY ||
    config.GHOST_STORAGE_ADAPTER_R2_SECRET_ACCESS_KEY;

  config.GHOST_STORAGE_ADAPTER_R2_CONTENT_PREFIX =
    process.env.GHOST_STORAGE_ADAPTER_R2_CONTENT_PREFIX ||
    config.GHOST_STORAGE_ADAPTER_R2_CONTENT_PREFIX ||
    '';

  config.GHOST_STORAGE_ADAPTER_R2_IMAGES_URL_PREFIX =
    process.env.GHOST_STORAGE_ADAPTER_R2_IMAGES_URL_PREFIX ||
    config.GHOST_STORAGE_ADAPTER_R2_IMAGES_URL_PREFIX ||
    '/content/images/';
  config.GHOST_STORAGE_ADAPTER_R2_MEDIA_URL_PREFIX =
    process.env.GHOST_STORAGE_ADAPTER_R2_MEDIA_URL_PREFIX ||
    config.GHOST_STORAGE_ADAPTER_R2_MEDIA_URL_PREFIX ||
    '/content/media/';
  config.GHOST_STORAGE_ADAPTER_R2_FILES_URL_PREFIX =
    process.env.GHOST_STORAGE_ADAPTER_R2_FILES_URL_PREFIX ||
    config.GHOST_STORAGE_ADAPTER_R2_FILES_URL_PREFIX ||
    '/content/files/';

  const responsiveImages: boolean | undefined = getBooleanFromEnv(
    'GHOST_STORAGE_ADAPTER_R2_RESPONSIVE_IMAGES'
  );
  config.GHOST_STORAGE_ADAPTER_R2_RESPONSIVE_IMAGES =
    responsiveImages ||
    config.GHOST_STORAGE_ADAPTER_R2_RESPONSIVE_IMAGES ||
    false;

  config.GHOST_STORAGE_ADAPTER_R2_RESIZE_WIDTHS =
    process.env.GHOST_STORAGE_ADAPTER_R2_RESIZE_WIDTHS ||
    config.GHOST_STORAGE_ADAPTER_R2_RESIZE_WIDTHS ||
    '300,600,1000,1600,400,750,960,1140,1200';

  const uuidName: boolean | undefined = getBooleanFromEnv(
    'GHOST_STORAGE_ADAPTER_R2_UUID_NAME'
  );
  config.GHOST_STORAGE_ADAPTER_R2_UUID_NAME =
    uuidName ?? (config.GHOST_STORAGE_ADAPTER_R2_UUID_NAME || false);

  let jpegQuality: number | undefined;
  if (process.env.GHOST_STORAGE_ADAPTER_R2_RESIZE_JPEG_QUALITY) {
    jpegQuality = parseInt(
      process.env.GHOST_STORAGE_ADAPTER_R2_RESIZE_JPEG_QUALITY
    );
  }
  config.GHOST_STORAGE_ADAPTER_R2_RESIZE_JPEG_QUALITY =
    jpegQuality || config.GHOST_STORAGE_ADAPTER_R2_RESIZE_JPEG_QUALITY || 80;

  const saveOriginal: boolean | undefined = getBooleanFromEnv(
    'GHOST_STORAGE_ADAPTER_R2_SAVE_ORIGINAL'
  );
  config.GHOST_STORAGE_ADAPTER_R2_SAVE_ORIGINAL =
    saveOriginal ?? (config.GHOST_STORAGE_ADAPTER_R2_SAVE_ORIGINAL || true);

  const ghostResize: boolean | undefined = getBooleanFromEnv(
    'GHOST_STORAGE_ADAPTER_R2_GHOST_RESIZE'
  );
  config.GHOST_STORAGE_ADAPTER_R2_GHOST_RESIZE =
    ghostResize ?? (config.GHOST_STORAGE_ADAPTER_R2_GHOST_RESIZE || true);

  const saveNameMetadata: boolean | undefined = getBooleanFromEnv(
    'GHOST_STORAGE_ADAPTER_R2_SAVE_ORIG_NAME_METADATA'
  );
  config.GHOST_STORAGE_ADAPTER_R2_SAVE_ORIG_NAME_METADATA =
    saveNameMetadata || false;

  config.GHOST_STORAGE_ADAPTER_R2_SAVE_ORIG_NAME_METADATA =
    saveNameMetadata ??
    (config.GHOST_STORAGE_ADAPTER_R2_SAVE_ORIG_NAME_METADATA || false);

  return config;
}

function checkConfig(config: Config) {
  if (!config.GHOST_STORAGE_ADAPTER_R2_DOMAIN) {
    throw new Error(
      'Environment/config variable "GHOST_STORAGE_ADAPTER_R2_DOMAIN" has not been set'
    );
  }
  if (!config.GHOST_STORAGE_ADAPTER_R2_BUCKET) {
    throw new Error(
      'Environment/config variable "GHOST_STORAGE_ADAPTER_R2_BUCKET" has not been set'
    );
  }
  if (!config.GHOST_STORAGE_ADAPTER_R2_ENDPOINT) {
    throw new Error(
      'Environment/config variable "GHOST_STORAGE_ADAPTER_R2_ENDPOINT" has not been set'
    );
  }
  if (!config.GHOST_STORAGE_ADAPTER_R2_ACCESS_KEY_ID) {
    throw new Error(
      'Environment/config variable "GHOST_STORAGE_ADAPTER_R2_ACCESS_KEY_ID" has not been set'
    );
  }
  if (!config.GHOST_STORAGE_ADAPTER_R2_SECRET_ACCESS_KEY) {
    throw new Error(
      'Environment/config variable "GHOST_STORAGE_ADAPTER_R2_SECRET_ACCESS_KEY" has not been set'
    );
  }
}

export default class CloudflareR2Adapter extends StorageBase {
  private S3: S3Client;
  private bucket: string;
  private pathPrefix: string;
  private domain: string;
  private storageType: StorageType = StorageType.Images;
  private imagesUrlPrefix: string;
  private mediaUrlPrefix: string;
  private filesUrlPrefix: string;
  private responsiveImages: boolean;
  private resizeWidths: number[];
  public saveRaw: unknown = undefined;
  private uuidName: boolean;
  private saveOriginal: boolean;
  private jpegQuality: number | undefined;
  private ghostResize: boolean;
  private contentPrefix: string;
  private saveOrigNameMetadata: boolean;

  constructor(config: Config = {}) {
    log.debug('Initialising ghost-cloudflare-r2 storage adapter');
    super();
    mergeConfigWithEnv(config);
    checkConfig(config);

    this.bucket = <string>config.GHOST_STORAGE_ADAPTER_R2_BUCKET;
    this.domain = <string>config.GHOST_STORAGE_ADAPTER_R2_DOMAIN;
    this.imagesUrlPrefix = <string>(
      config.GHOST_STORAGE_ADAPTER_R2_IMAGES_URL_PREFIX
    );
    this.mediaUrlPrefix = <string>(
      config.GHOST_STORAGE_ADAPTER_R2_MEDIA_URL_PREFIX
    );
    this.filesUrlPrefix = <string>(
      config.GHOST_STORAGE_ADAPTER_R2_FILES_URL_PREFIX
    );
    this.responsiveImages = <boolean>(
      config.GHOST_STORAGE_ADAPTER_R2_RESPONSIVE_IMAGES
    );

    this.resizeWidths = (<string>config.GHOST_STORAGE_ADAPTER_R2_RESIZE_WIDTHS)
      .split(',')
      .map(w => parseInt(w));

    if (config.GHOST_STORAGE_ADAPTER_R2_RESPONSIVE_IMAGES === true) {
      // Ghost checks if a 'saveRaw' function exists on the storage adapter,
      // if it exists, the theme will generate srcset attribute in the HTML.
      this.saveRaw = function () {};
    }

    if (config.storage_type_images === true) {
      this.storageType = StorageType.Images;
      this.pathPrefix = <string>(
        config.GHOST_STORAGE_ADAPTER_R2_IMAGES_URL_PREFIX
      );
    } else if (config.storage_type_media === true) {
      this.storageType = StorageType.Media;
      this.pathPrefix = <string>(
        config.GHOST_STORAGE_ADAPTER_R2_MEDIA_URL_PREFIX
      );
    } else if (config.storage_type_files === true) {
      this.storageType = StorageType.Files;
      this.pathPrefix = <string>(
        config.GHOST_STORAGE_ADAPTER_R2_FILES_URL_PREFIX
      );
    } else {
      this.storageType = StorageType.Images;
      this.pathPrefix = this.imagesUrlPrefix;
    }
    this.contentPrefix = config.GHOST_STORAGE_ADAPTER_R2_CONTENT_PREFIX || '';
    this.pathPrefix = this.contentPrefix + this.pathPrefix;

    this.uuidName = <boolean>config.GHOST_STORAGE_ADAPTER_R2_UUID_NAME;
    this.saveOriginal = <boolean>config.GHOST_STORAGE_ADAPTER_R2_SAVE_ORIGINAL;
    this.ghostResize = <boolean>config.GHOST_STORAGE_ADAPTER_R2_GHOST_RESIZE;
    this.saveOrigNameMetadata = <boolean>(
      config.GHOST_STORAGE_ADAPTER_R2_SAVE_ORIG_NAME_METADATA
    );

    log.info(
      'Cloudflare R2 Storage Adapter: handling',
      StorageType[this.storageType],
      'at',
      this.pathPrefix
    );

    this.S3 = new S3Client({
      region: 'auto',
      endpoint: <string>config.GHOST_STORAGE_ADAPTER_R2_ENDPOINT,
      credentials: {
        accessKeyId: <string>config.GHOST_STORAGE_ADAPTER_R2_ACCESS_KEY_ID,
        secretAccessKey: <string>(
          config.GHOST_STORAGE_ADAPTER_R2_SECRET_ACCESS_KEY
        ),
      },
    });
    log.debug('Initialisation done');
  }

  delete(fileName: string, targetDir?: string): Promise<boolean> {
    log.debug('delete():', 'filename:', fileName, 'targetDir:', targetDir);
    log.error('Cloudflare R2 Storage Adapter: delete() is not implemented');
    return Promise.resolve(false);
  }

  exists(fileName: string, targetDir?: string): Promise<boolean> {
    log.info('exists():', 'fileName:', fileName, 'targetDir:', targetDir);

    let targetPath: string;

    if (targetDir === undefined) {
      targetPath = fileName;
    } else {
      targetPath = path.join(targetDir, fileName);
    }

    return new Promise((resolve, reject) => {
      this.S3.send(
        new HeadObjectCommand({
          Bucket: this.bucket,
          Key: stripLeadingSlash(targetPath),
        })
      )
        .then(
          value => {
            if (value.$metadata.httpStatusCode === 200) {
              resolve(true);
            } else {
              resolve(false);
            }
          },
          reason => {
            resolve(false);
          }
        )
        .catch(reason => {
          log.debug(reason);
          reject(reason);
        });
    });
  }

  read(options?: StorageBase.ReadOptions): Promise<Buffer> {
    log.debug('Cloudflare R2 Storage Adapter: read():', 'options:', options);

    return new Promise((resolve, reject) => {
      if (options === undefined) {
        reject(
          'Cloudflare R2 Storage Adapter: read(): argument "options" is undefined'
        );
        return;
      }

      if (options?.path === undefined) {
        reject(
          'Cloudflare R2 Storage Adapter: read(): argument "options.path" is undefined'
        );
        return;
      }

      const r2Path = stripLeadingSlash(
        path.join(this.pathPrefix, options?.path)
      );

      this.S3.send(
        new GetObjectCommand({
          Bucket: this.bucket,
          Key: r2Path,
        })
      )
        .then(
          value => {
            value.Body?.transformToByteArray()
              .then(
                (value: Uint8Array) => {
                  resolve(Buffer.from(value));
                },
                (reason: unknown) => {
                  reject(reason);
                }
              )
              .catch((err: unknown) => {
                reject(err);
              });
          },
          reason => {
            reject(reason);
          }
        )
        .catch(err => {
          reject(err);
        });
    });
  }

  saveResizedImages(
    fileInfo: FileInfo,
    fileBuffer: Buffer,
    originalUuid: string | null,
    isImport: boolean
  ): Promise<boolean> {
    log.info(
      'Cloudflare R2 Storage Adapter: saveResizedImages(): fileInfo:',
      fileInfo
    );

    return new Promise((resolve, reject) => {
      Promise.all(
        this.resizeWidths.map(width => {
          let directory = this.getTargetDir(
            `${stripEndingSlash(this.pathPrefix)}/size/w${width}`
          );

          if (isImport) {
            // transform /content/images/2022/12/image.jpg
            // Into /content_prefix/content/images/size/w_x/2022/12
            if (fileInfo.newPath === undefined) {
              reject(`Could not determine newPath for image ${fileInfo.path}`);
              return;
            }

            const oldDir = stripLeadingSlash(fileInfo.newPath)
              .split('/')
              .slice(0, -1);
            if (oldDir === undefined) {
              reject(`Could not determine newPath for image ${fileInfo.path}`);
              return;
            }
            // WARNING: assume old path structure was in the format /content/images/size/wX/image.jpg
            directory = `${this.contentPrefix}/${oldDir[0]}/${oldDir[1]}/size/w${width}/${oldDir[2]}/${oldDir[3]}`;
          }

          return Promise.all([
            this.getUniqueFileName(fileInfo, directory, originalUuid),
            this.jpegQuality && fileInfo.type === 'image/jpeg'
              ? sharp(fileBuffer)
                  .resize({width: width})
                  .jpeg({quality: this.jpegQuality})
                  .toBuffer()
              : sharp(fileBuffer).resize({width: width}).toBuffer(),
          ])
            .then(([filePathR2, resizedBuffer]) => {
              log.debug(
                'Cloudflare R2 Storage Adapter: saveResizedImages(): saving',
                filePathR2
              );

              return this.S3.send(
                new PutObjectCommand({
                  Bucket: this.bucket,
                  Body: resizedBuffer,
                  ContentType: fileInfo.type,
                  CacheControl: `max-age=${30 * 24 * 60 * 60}`,
                  Key: stripLeadingSlash(filePathR2),
                })
              ).then(() => {
                log.info('Saved', filePathR2);
              });
            })
            .catch(reason => {
              reject(reason);
            });
        })
      )
        .then(() => {
          log.debug('Finished saving resized images for', fileInfo.name);
          resolve(true);
        })
        .catch(reason => {
          reject(reason);
        });
    });
  }

  getUniqueFileName(
    fileInfo: FileInfo,
    targetDir: string,
    uuid?: string | null
  ): string {
    if (this.storageType === StorageType.Files) {
      return super.getUniqueFileName(fileInfo, targetDir);
    }

    if (uuid) {
      return path.join(targetDir, uuid + fileInfo.ext);
    } else {
      return super.getUniqueFileName(fileInfo, targetDir);
    }
  }

  isOriginalImage(fileInfo: FileInfo): boolean {
    return !fileInfo.path.endsWith('_processed');
  }

  save(
    fileInfo: FileInfo,
    targetDir?: string,
    forceUuid?: string
  ): Promise<string> {
    log.info(
      'Cloudflare R2 Storage Adapter: save():',
      'fileInfo:',
      fileInfo,
      'targetDir:',
      targetDir
    );

    let isImport = false;
    if (targetDir) {
      log.info('Cloudflare R2 Storage Adapter: save(): Detected import.');
      isImport = true;
      fileInfo.name = path.basename(fileInfo.name);
      fileInfo.ext = path.extname(fileInfo.name);
    }

    let directory = this.getTargetDir(this.pathPrefix);

    // getTargetDir adds year/month. For import we want the original path
    if (isImport) {
      directory =
        this.contentPrefix +
        fileInfo.newPath?.split('/').slice(0, -1).join('/');
    }

    return new Promise((resolve, reject) => {
      if (
        !this.saveOriginal &&
        this.isOriginalImage(fileInfo) &&
        this.ghostResize &&
        !isImport
      ) {
        log.info(
          'Cloudflare R2 Storage Adapter: save(): discarding original: ',
          fileInfo.name
        );
        // Not sure if the URL for the original image is used.
        // Should only be used if imageOptimization__resize is true since then the original (*_o.jpg) is not used.
        resolve('');
        return;
      }

      let uuid: string | null = null;
      if (this.uuidName && !isImport) {
        uuid = forceUuid || uuidv4();
      }

      Promise.all([
        this.getUniqueFileName(fileInfo, directory, uuid),
        readFileAsync(fileInfo.path),
      ])
        .then(([filePathR2, fileBuffer]) => {
          if (fileInfo.type === '' || fileInfo.type === undefined) {
            const mimeType = mime.getType(fileInfo.path);
            if (mimeType) {
              log.debug('Detected mimeType:', mimeType);
              fileInfo.type = mimeType;
            }
          }
          log.debug(
            'Cloudflare R2 Storage Adapter: save(): saving',
            filePathR2
          );

          let metadata = {};
          if (this.saveOrigNameMetadata) {
            if (isImport) {
              metadata = {original_name: path.basename(fileInfo.name)};
            } else {
              if (fileInfo.originalname === undefined) {
                throw new Error(
                  'save(): originalname is not defined for non import, could not save original name metadata'
                );
              }
              metadata = {original_name: path.basename(fileInfo.originalname)};
            }
          }
          this.S3.send(
            new PutObjectCommand({
              Bucket: this.bucket,
              Body: fileBuffer,
              ContentType: fileInfo.type,
              CacheControl: `max-age=${30 * 24 * 60 * 60}`,
              Key: stripLeadingSlash(filePathR2),
              Metadata: metadata,
            })
          ).then(
            () => {
              log.info('Saved', filePathR2);
              if (
                ((this.ghostResize && !this.isOriginalImage(fileInfo)) ||
                  (!this.ghostResize && this.isOriginalImage(fileInfo)) ||
                  isImport) &&
                this.responsiveImages &&
                this.storageType === StorageType.Images
              ) {
                log.info('Generating different image sizes...');
                this.saveResizedImages(fileInfo, fileBuffer, uuid, isImport)
                  .then(() => {
                    log.info('Generating different image sizes... Done');
                    resolve(`${this.domain}/${stripLeadingSlash(filePathR2)}`);
                  })
                  .catch(reason => {
                    reject(reason);
                  });
              } else {
                resolve(`${this.domain}/${stripLeadingSlash(filePathR2)}`);
              }
            },
            reason => {
              reject(reason);
            }
          );
        })
        .catch(err => {
          log.debug(err);
          reject(err);
        });
    });
  }

  serve(): Handler {
    return (req, res, next) => {
      next();
    };
  }
}

module.exports = CloudflareR2Adapter;
